### PR TITLE
post 01/23 hackathon improvements in the expectation contrib tooling and documentation

### PR DESCRIPTION
**Pull request by eugmandel**
### Changes proposed in this pull request:
- Expectation.run_disagnostics was displaying an unclear error when sqlalchemy was not installed.
  Here is how this PR addresses it:
  1. Expectation.run_disagnostics was modified to catch exceptions raised during the running of the tests and add them to the report
  2. build_test_backends_list was cleaned up to support this: clear arguments for which backends are requested and raise an error if a requested backend cannot be created.

- Renamed success key under test_report to test_passed - this was flagged as confusing
- Added an acceptance checklist for contrib PRs

Previous Design Review notes:
-
-
-

### CODE REVIEW

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+================================================
+
+This acceptance checklist applies to "contrib" Pull Requests - PRs that make modifications only in ``/contrib`` directory of the repo which contains community contributed Expectations.
+
+
+Both contributors and reviewers should follow this checklist:
+
+* Scope of the PR.
+    * A PR may introduce a new Expectation. The minimal acceptable version of a new Expectation is a module that contains a new class with the name of the Expectation, a Docstring describing the behavior, and examples/tests. The class may leave all methods unimplemented. Metric classes required by the new Expectation may be left undefined/unimplemented. This minimal contribution can be viewed as a "RFE" - Request for Expectation.
+    * If the PR modifies existing Expectations and/or Metrics, check that the changes are an improvement and are not destructive. The contributor should not remove someone else' prior work without a clear reason.
```
Comment (jcampbell) [inline docs/contributing/contrib_pull_request_acceptance_checklist.rst]:
```suggestion
    * If the PR modifies existing Expectations and/or Metrics, check that the changes are an improvement and are not destructive. The contributor should not remove someone else's prior work without a clear reason.
```

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+* Basic class/module structure and naming of new Expectation and Metric classes
+    * One new Expectation class per module.
```
Comment (jcampbell) [inline docs/contributing/contrib_pull_request_acceptance_checklist.rst]:
```suggestion
    * One new Expectation class per module (file).
```

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    * The modules are placed in a subdirectory of `/contrib/experimental/expectations` that makes sense for the new Expectation
+    * The new Expectation classes follow CamelCase convention. Module names are snake_case versions of the Expectation class names.
+    * Metric classes that implement the new Metrics the Expectation depends on are implemented in the same module.
```
Comment (jcampbell) [inline docs/contributing/contrib_pull_request_acceptance_checklist.rst:20]: (when) will we allow metrics in their own module?

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    * Expectation class names summarize the nature of the expectation - this is a judgement call.
```
Comment (jcampbell) [inline docs/contributing/contrib_pull_request_acceptance_checklist.rst]:
```suggestion
    * Expectation class names summarize the nature of the expectation.
```

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    * New Expectation and Metric classes extend the appropriate core classes (e.g., new column map expectations extend ColumnMapExpectation)
+
+* Docstring
+    * Docstring must contain a legible description of the functionality of the new Expectations.
+
+* Library metatada
+    * this is the information required to add an Expectation to the Public Expectation Gallery. We expect the contributor to fill out this section, but we will "pick up the slack" if needed. The module containing a new Expectation must have a comment that looks as follows:
```
Comment (jcampbell) [inline docs/contributing/contrib_pull_request_acceptance_checklist.rst]:
```suggestion
    * The `library_metadata` is the information required to add an Expectation to the Public Expectation Gallery. Contributors, please complete as much of the section as possible. Reviewers, please complete any remaining fields to avoid extra asks to contributors. Adapt the metadata example below as appropriate for the new contribution.
```

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            "contributors": [ # Github handles for all contributors to this Expectation.
+        #         "@your_name_here", # Don't forget to add your github handle here!
+            ],
+            "package": "experimental_expectations",
+        }
+
+
+
+* Examples/tests
+    * The module that contains a wew Expectations must have ``examples`` variable defined. The variable's value is a structure that defines inputs and outputs that serve both as examples and as tests. At least one positive and at least one negative example must be present. Below is an example of this structure for an imaginary expectation ``expect_column_values_to_begin_with_specific_letter``:
```
Comment (jcampbell) [inline docs/contributing/contrib_pull_request_acceptance_checklist.rst]:
```suggestion
    * Any module that contains a new Expectation must have ``examples`` defined. The examples variable defines data, inputs, and outputs that serve both as examples and as tests. At least one positive and at least one negative example must be present. Below is an example of this structure for an imaginary expectation ``expect_column_values_to_begin_with_specific_letter``:
```

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                {
+                    "title": "maintest",
+                    "exact_match_out": False,
+                    "in": {"column": "myvalues", "letter": "a", "mostly": 0.4},
+                    "out": {
+                        "success": True,
+                        "unexpected_index_list": [1, 4],
+                        "unexpected_list": ["pear", "mouse"]
+                    },
+                }
```
Comment (jcampbell) [inline docs/contributing/contrib_pull_request_acceptance_checklist.rst]: This needs a negative example as required by the "both positive and negative" criteria.

```suggestion
                {
                    "title": "basic_positive_test",
                    "exact_match_out": False,
                    "in": {"column": "myvalues", "letter": "a", "mostly": 0.4},
                    "out": {
                        "success": True,
                        "unexpected_index_list": [1, 4],
                        "unexpected_list": ["pear", "mouse"]
                    },
                    "gallery_example": True
                },
                {
                    "title": "basic_negative_test",
                    "exact_match_out": False,
                    "in": {"column": "myvalues", "letter": "p", "mostly": 0.4},
                    "out": {
                        "success": False,
                        "unexpected_index_list": [0, 2, 3, 4, 5],
                        "unexpected_list": ["apple", "acyclical graph", "aardvardk", "mouse", "angular"],
                    },
                    "gallery_example": True
                }
```

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            ],
+        }]
+
+    * Test titles must contain no spaces so that they can be individually selected by pytest.
+
+    * If a test should be rendered as an example, it should be marked up with ``gallery_example: True``
```
Comment (jcampbell) [inline docs/contributing/contrib_pull_request_acceptance_checklist.rst]: I added to the above...please confirm.

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    * If new Expectation class in the PR goes beyond "RFE" (it does not just define the name and the Docstring of the new Expectation, but has elements of implementation), the tests included in the module must pass for the PR to be approved.
```
Comment (jcampbell) [inline docs/contributing/contrib_pull_request_acceptance_checklist.rst]: If there is no implementation, but there are tests, what currently happens?

```suggestion
    * The tests included in the module must pass for the PR to be approved.
```

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+ * If the PR contains changes to existing renderers (or new renderers), the wording must "make sense" - this is a judgement call.
```
Comment (jcampbell) [inline docs/contributing/contrib_pull_request_acceptance_checklist.rst]:
```suggestion
 * Verify that changes to existing renderers (or new renderers) make sense--the language needs to accurately describe the semantics and behavior of the Expectation. 
```

File: docs/guides/how_to_guides/creating_and_editing_expectations/how_to_contribute_a_new_expectation_to_great_expectations.rst [hunk 11]
```diff
@@ -160,7 +160,7 @@ Steps
 
     Each example is a dictionary with two keys:
 
-    * ``data``: defines the input data of the example as a table/data frame. In this example the table has one column named "mostly_threes" with 10 rows.
+    * ``data``: defines the input data of the example as a table/data frame. In this example the table has one column named "mostly_threes" with 10 rows. If you define multiple columns, make sure that they have the same number of rows.
```
Comment (jcampbell) [inline docs/guides/how_to_guides/creating_and_editing_expectations/how_to_contribute_a_new_expectation_to_great_expectations.rst]:
```
    * ``data``: defines the input data of the example as a table/data frame. In this example the table has one column named "mostly_threes" with 10 rows. If you define multiple columns, make sure that they have the same number of rows. If possible, include test data and tests that includes null values (``None`` in the python test definition).
```

File: great_expectations/expectations/expectation.py [hunk 12]
```diff
@@ -924,15 +929,15 @@ def _get_test_results(
                     {
                         "test title": exp_test["test"]["title"],
                         "backend": exp_test["backend"],
-                        "success": "true",
+                        "test_passed": "true",
```
Comment (jcampbell) [inline great_expectations/expectations/expectation.py:932]: <3 I hadn't thought of that ambiguity before!!

File: great_expectations/expectations/self_check_util.py [hunk 13]
```diff
@@ -1650,24 +1650,38 @@ def is_valid(self):
             return self._is_valid
 
 
-def build_test_backends_list():
-    test_backends = ["pandas"]
-    no_spark = True
-    if not no_spark:
+def build_test_backends_list(
```
Comment (jcampbell) [inline great_expectations/expectations/self_check_util.py:1653]: Can we do a quick pycharm refactor to rename "backends" to "execution_engines" now that we've standardized on that name? If it's more than a quick operation, no need.

**Review by jcampbell — CHANGES_REQUESTED** [review]
Looks awesome -- inline changes mostly for clarity/typos.

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst (hunk 3 above)
Comment (eugmandel) [reply docs/contributing/contrib_pull_request_acceptance_checklist.rst:20]: I believe we will, but not in the current iteration.

File: docs/guides/how_to_guides/creating_and_editing_expectations/how_to_contribute_a_new_expectation_to_great_expectations.rst (hunk 11 above)
Comment (eugmandel) [reply docs/guides/how_to_guides/creating_and_editing_expectations/how_to_contribute_a_new_expectation_to_great_expectations.rst]: Thank you! -pushed.

File: great_expectations/expectations/self_check_util.py (hunk 13 above)
Comment (eugmandel) [reply great_expectations/expectations/self_check_util.py:1653]: Let's do this in a separate PR.

**Review by jcampbell — APPROVED** [review]

File: docs/contributing/contrib_pull_request_acceptance_checklist.rst (hunk 8 above)
Comment (eugmandel) [reply docs/contributing/contrib_pull_request_acceptance_checklist.rst]: confirmed  + renamed to "include_in_gallery" to be consistent with the code and the how-to guide.
